src/modm/platform/adc/stm32f3/adc_interrupt.cpp.in

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright (c) 2009-2010, Fabian Greif
+ * Copyright (c) 2009-2010, Martin Rosekeit
+ * Copyright (c) 2012, 2014-2015, 2017, Niklas Hauser
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#include "adc_interrupt_{{ id }}.hpp"
+#include <modm/architecture/interface/interrupt.hpp>
+// ----------------------------------------------------------------------------
+modm::platform::AdcInterrupt{{ id }}::Handler
+modm::platform::AdcInterrupt{{ id }}::handler(modm::dummy);
+
+%% if id not in shared_irq_ids
+MODM_ISR(ADC{{ id }})
+{
+    if (auto flags = modm::platform::AdcInterrupt{{ id }}::getInterruptFlags()) {
+	modm::platform::Adc{{ id }}::acknowledgeInterruptFlag(flags);
+        modm::platform::AdcInterrupt{{ id }}::handler();
```

Comment:
There is a discrepancy here to [STM32F1/F4 Adc interrupt handlers](https://github.com/modm-io/modm/blob/develop/src/modm/platform/adc/stm32/adc_interrupt.cpp.in). There it is the attached handler's responsibility to acknowledge the interrupt flags. Here it is done within the ISR that calls the handler.
After a superficial look I didn't see a reason to not unify the behavior of both drivers. I have never used the F3 Adc, so I don't know the specifics here. We should try to keep the drivers consistent unless there is a relevant technical reason not to do so.

@XDjackieXD Do you know why this is necessary for F3 to make it work with AdcSampler?  

And why isn't it for the F1/F4 driver (assuming it works with `AdcSampler` ...) ?
The EndOfConversion flag is implicitly cleared by reading the data register for both Adc IPs. This is always done inside the `AdcSampler` handler. `Adc::startConversion()` is called in most code paths but not all. For F3 starting a conversion clears some flags (there is a TODO in the code), the other driver clears all.

I could imagine that a handler would actually want to know which condition led to the interrupt being triggered in some applications. When the flag in cleared before the handler is called that could be difficult. Maybe we should fix the problem with clearing the flags in AdcSampler? What do you think? @XDjackieXD @rleh @salkinium

Reply:
Acking the flags in sample probably is the cleaner solution. yes.
This was primarily my solution too "there seem to be more flags set than sampler ACKed" and it worked so I did not change it. This is the reason why this PR is still a draft :P

Reply:
I wrote AdcSampler for the AVR, before we even had any common Adc interface, so it should definitely get updated.